server/routes/serviceProviderReferrals/serviceProviderReferralsController.ts

```diff
@@ -1483,6 +1485,32 @@ export default class ServiceProviderReferralsController {
     ControllerUtils.renderWithLayout(res, view, serviceUser)
   }
 
+  async viewEndOfServiceReport(req: Request, res: Response): Promise<void> {
+    const { accessToken } = res.locals.user.token
+    const endOfServiceReport = await this.interventionsService.getEndOfServiceReport(accessToken, req.params.id)
+    const referral = await this.interventionsService.getSentReferral(accessToken, endOfServiceReport.referralId)
+    const intervention = await this.interventionsService.getIntervention(accessToken, referral.referral.interventionId)
+    const serviceCategories = intervention.serviceCategories.filter(serviceCategory =>
+      referral.referral.serviceCategoryIds.some(serviceCategoryId => serviceCategoryId === serviceCategory.id)
+    )
+    if (serviceCategories.length !== referral.referral.serviceCategoryIds.length) {
+      throw new Error('Expected service categories are missing in intervention')
+    }
+
+    if (endOfServiceReport.submittedAt === null) {
+      throw new Error(
+        'You cannot view an end of service report that has not yet been submitted. Please submit the end of service report before trying to view it.'
```

Comment:
It's good to add additional protection but I am wondering how would the user ever have access to this page?

The link to this page would only be shown when the EOSR is submitted and the UUID is impossible to guess?

Reply:
It's definitely very unlikely - I thought best to err on the side of caution though to make it a bit more future-proof in case e.g. we decide to allow people to edit EOSRs.